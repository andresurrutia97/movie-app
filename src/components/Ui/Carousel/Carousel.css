.slider {
  display: flex;
  position: relative;
}

.slider-title {
  margin-left: 20px;
  font-weight: 600;
  font-size: 18px;
  margin-bottom: 10px;
}

.slider__container {
  display: flex;
  margin: 0 12px;
  transition: transform 300ms ease 100ms;
  z-index: 3;
  width: 100%;
}

@media (min-width: 768px) {
  .slider__container {
    padding: 0 76px;
    margin: 0;
  }

  .slider-title {
    margin-left: 80px;
  }

  .slider:not(slider--open) .slider-item:hover {
    transform: scale(1.2) !important;
  }

  .slider:not(slider--open):hover .slider-item {
    transform: translateX(-10%);
  }

  .slider:not(slider--open) .slider-item:hover ~ .slider-item {
    transform: translateX(10%);
  }
}
