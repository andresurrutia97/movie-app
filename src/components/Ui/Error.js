import React from 'react';

const Error = () => (
  <p className="w-full py-5 text-center">Something went wrong</p>
);

export default Error;
