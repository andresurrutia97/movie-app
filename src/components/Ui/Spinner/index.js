import React from 'react';

import './index.css';

const Loading = () => <div className="spinner" />;

export default Loading;
