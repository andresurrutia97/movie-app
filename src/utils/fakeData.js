export const moviesData = {
  page: 1,
  results: [
    {
      poster_path: '/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg',
      adult: false,
      overview:
        'From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the United States government, undertaking high-risk black ops missions in exchange for commuted prison sentences.',
      release_date: '2016-08-03',
      genre_ids: [14, 28, 80],
      id: 297761,
      original_title: 'Suicide Squad',
      original_language: 'en',
      title: 'Suicide Squad',
      backdrop_path: '/ndlQ2Cuc3cjTL7lTynw6I4boP4S.jpg',
      popularity: 48.261451,
      vote_count: 1466,
      video: false,
      vote_average: 5.91,
    },
    {
      poster_path: '/lFSSLTlFozwpaGlO31OoUeirBgQ.jpg',
      adult: false,
      overview:
        'The most dangerous former operative of the CIA is drawn out of hiding to uncover hidden truths about his past.',
      release_date: '2016-07-27',
      genre_ids: [28, 53],
      id: 324668,
      original_title: 'Jason Bourne',
      original_language: 'en',
      title: 'Jason Bourne',
      backdrop_path: '/AoT2YrJUJlg5vKE3iMOLvHlTd3m.jpg',
      popularity: 30.690177,
      vote_count: 649,
      video: false,
      vote_average: 5.25,
    },
    {
      poster_path: '/hU0E130tsGdsYa4K9lc3Xrn5Wyt.jpg',
      adult: false,
      overview:
        'One year after outwitting the FBI and winning the publicâ€™s adulation with their mind-bending spectacles, the Four Horsemen resurface only to find themselves face to face with a new enemy who enlists them to pull off their most dangerous heist yet.',
      release_date: '2016-06-02',
      genre_ids: [28, 12, 35, 80, 9648, 53],
      id: 291805,
      original_title: 'Now You See Me 2',
      original_language: 'en',
      title: 'Now You See Me 2',
      backdrop_path: '/zrAO2OOa6s6dQMQ7zsUbDyIBrAP.jpg',
      popularity: 29.737342,
      vote_count: 684,
      video: false,
      vote_average: 6.64,
    },
    {
      poster_path: '/h28t2JNNGrZx0fIuAw8aHQFhIxR.jpg',
      adult: false,
      overview:
        'A recently cheated on married woman falls for a younger man who has moved in next door, but their torrid affair soon takes a dangerous turn.',
      release_date: '2015-01-23',
      genre_ids: [53],
      id: 241251,
      original_title: 'The Boy Next Door',
      original_language: 'en',
      title: 'The Boy Next Door',
      backdrop_path: '/vj4IhmH4HCMZYYjTMiYBybTWR5o.jpg',
      popularity: 22.279864,
      vote_count: 628,
      video: false,
      vote_average: 4.13,
    },
    {
      poster_path: '/vOipe2myi26UDwP978hsYOrnUWC.jpg',
      adult: false,
      overview:
        'An orphan boy is raised in the Jungle with the help of a pack of wolves, a bear and a black panther.',
      release_date: '2016-04-07',
      genre_ids: [12, 18, 14],
      id: 278927,
      original_title: 'The Jungle Book',
      original_language: 'en',
      title: 'The Jungle Book',
      backdrop_path: '/eIOTsGg9FCVrBc4r2nXaV61JF4F.jpg',
      popularity: 21.104822,
      vote_count: 1085,
      video: false,
      vote_average: 6.42,
    },
  ],
  total_results: 19629,
  total_pages: 982,
};

export const movieInfo = {
  adult: false,
  backdrop_path: '/fCayJrkfRaCRCTh8GqN30f8oyQF.jpg',
  belongs_to_collection: null,
  budget: 63000000,
  genres: [
    {
      id: 18,
      name: 'Drama',
    },
  ],
  homepage: '',
  id: 550,
  imdb_id: 'tt0137523',
  original_language: 'en',
  original_title: 'Fight Club',
  overview:
    'A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground "fight clubs" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.',
  popularity: 0.5,
  poster_path: null,
  production_companies: [
    {
      id: 508,
      logo_path: '/7PzJdsLGlR7oW4J0J5Xcd0pHGRg.png',
      name: 'Regency Enterprises',
      origin_country: 'US',
    },
    {
      id: 711,
      logo_path: null,
      name: 'Fox 2000 Pictures',
      origin_country: '',
    },
    {
      id: 20555,
      logo_path: null,
      name: 'Taurus Film',
      origin_country: '',
    },
    {
      id: 54050,
      logo_path: null,
      name: 'Linson Films',
      origin_country: '',
    },
    {
      id: 54051,
      logo_path: null,
      name: 'Atman Entertainment',
      origin_country: '',
    },
    {
      id: 54052,
      logo_path: null,
      name: 'Knickerbocker Films',
      origin_country: '',
    },
    {
      id: 25,
      logo_path: '/qZCc1lty5FzX30aOCVRBLzaVmcp.png',
      name: '20th Century Fox',
      origin_country: 'US',
    },
  ],
  production_countries: [
    {
      iso_3166_1: 'US',
      name: 'United States of America',
    },
  ],
  release_date: '1999-10-12',
  revenue: 100853753,
  runtime: 139,
  spoken_languages: [
    {
      iso_639_1: 'en',
      name: 'English',
    },
  ],
  status: 'Released',
  tagline:
    "How much can you know about yourself if you've never been in a fight?",
  title: 'Fight Club',
  video: false,
  vote_average: 7.8,
  vote_count: 3439,
};
