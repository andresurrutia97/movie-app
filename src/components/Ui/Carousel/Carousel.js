import React, { useState } from 'react';
import PropTypes from 'prop-types';

import './Carousel.css';
import useSliding from './useSliding';
import useSizeElement from './useSizeElement';
import SliderContext from './Context';
import SliderWrapper from './SliderWrapper';
import SlideButton from './SlideButton';
import Content from './Content';

const Carousel = ({ children, title }) => {
  const [currentSlide, setCurrentSlide] = useState(null);
  const { width, elementRef } = useSizeElement();
  const {
    handlePrev,
    handleNext,
    slideProps,
    containerRef,
    hasNext,
    hasPrev,
  } = useSliding(width, React.Children.count(children));

  /**
   * handleSelect() set the current slide to the
   * one that was selected by de user
   */
  const handleSelect = (movie) => {
    setCurrentSlide(movie);
  };

  /**
   * handleClose() set the current slide to null
   */
  const handleClose = () => {
    setCurrentSlide(null);
  };

  const contextValue = {
    onSelectSlide: handleSelect,
    onCloseSlide: handleClose,
    elementRef,
    currentSlide,
  };

  let slider = ['slider'];
  if (currentSlide != null) {
    slider = [...slider, 'slider--open'];
  }
  return (
    <SliderContext.Provider value={contextValue}>
      <SliderWrapper>
        <div className="slider-title">{title}</div>
        <div className={slider}>
          <div ref={containerRef} className="slider__container" {...slideProps}>
            {children}
          </div>
        </div>
        {hasPrev && <SlideButton onClick={handlePrev} type="prev" />}
        {hasNext && <SlideButton onClick={handleNext} type="next" />}
      </SliderWrapper>
      {currentSlide && <Content movie={currentSlide} onClose={handleClose} />}
    </SliderContext.Provider>
  );
};

Carousel.propTypes = {
  children: PropTypes.node,
  title: PropTypes.string,
};

Carousel.defaultProps = {
  children: null,
  title: '',
};

export default Carousel;
