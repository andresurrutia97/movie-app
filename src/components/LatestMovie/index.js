import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import LatestMovieContent from './LatestMovieContent';
import Button from '../Ui/Button';
import Title from '../Ui/MovieTitle';
import MovieGenres from '../Ui/MovieGenres';
import Runtime from '../Ui/MovieRuntime';
import Overview from '../Ui/MovieOverview';
import ImgBackground from './Background';
import Shadow from './Shadow';
import { goToMovie } from '../../utils/navigation';

const LatestMovie = ({ movie, history }) => {
  return (
    <div className="w-full relative">
      <ImgBackground img={movie.backdrop_path} />
      <Shadow />
      <LatestMovieContent>
        <div className="flex flex-col items-center md:w-full md:items-start">
          <Title title={movie.title} />
          <MovieGenres genres={movie.genres} />
          <Runtime runtime={movie.runtime} />
          <Overview overview={movie.overview} />
        </div>
        <div className="mt-1 w-24 md:mt-12 md:w-64">
          <Button onClick={() => goToMovie(history, movie.id)}>Watch</Button>
        </div>
      </LatestMovieContent>
    </div>
  );
};

LatestMovie.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number,
    backdrop_path: PropTypes.string,
    title: PropTypes.string,
    genres: PropTypes.array,
    runtime: PropTypes.number,
    overview: PropTypes.string,
  }).isRequired,
  history: PropTypes.any.isRequired,
};

export default withRouter(LatestMovie);
