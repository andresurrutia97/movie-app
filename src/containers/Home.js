import React from 'react';
import { useDispatch, useSelector } from '../utils/reduxHooks';

import HomeView from '../views/Home';
import {
  fetchPopularMovies,
  fetchTopRatedMovies,
} from '../store/actions/MoviesActions';
import { fetchFavoritesListIfNeeded } from '../store/actions/FavoritesActions';
import { fetchSeenListIfNeeded } from '../store/actions/SeenMoviesActions';

const Home = () => {
  const {
    popularMovies,
    loadingPopularMovies,
    errorPopularMovies,
    topRatedMovies,
    loadingTopRatedMovies,
    errorTopRatedMovies,
    latestMovie,
    loadingLatestMovie,
    errorLatestMovie,
    loadingFavList,
    errorFavList,
    favorites,
    loadingFavs,
    errorFavs,
    seen,
    loadingSeen,
    errorSeen,
  } = useSelector((state) => ({
    popularMovies: state.movies.popularMovies,
    loadingPopularMovies: state.movies.loadingPopularMovies,
    errorPopularMovies: state.movies.errorPopularMovies,
    topRatedMovies: state.movies.topRatedMovies,
    loadingTopRatedMovies: state.movies.loadingTopRatedMovies,
    errorTopRatedMovies: state.movies.errorTopRatedMovies,
    latestMovie: state.movies.latestMovie,
    loadingLatestMovie: state.movies.loadingLatestMovie,
    errorLatestMovie: state.movies.errorLatestMovie,
    loadingFavList: state.favorites.loadingFavList,
    errorFavList: state.favorites.errorFavList,
    favorites: state.favorites.favorites,
    loadingFavs: state.favorites.loadingFavs,
    errorFavs: state.favorites.errorFavs,
    seen: state.seen.seen,
    loadingSeen: state.seen.loadingSeen,
    errorSeen: state.seen.errorSeen,
  }));

  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(fetchPopularMovies());
    dispatch(fetchTopRatedMovies());
    dispatch(fetchFavoritesListIfNeeded());
    dispatch(fetchSeenListIfNeeded());
  }, [dispatch]);

  return (
    <HomeView
      popularMovies={popularMovies}
      loadingPopularMovies={loadingPopularMovies}
      errorPopular={errorPopularMovies}
      topRatedMovies={topRatedMovies}
      loadingTopRatedMovies={loadingTopRatedMovies}
      errorTopRated={errorTopRatedMovies}
      latestMovie={latestMovie}
      loadingLatestMovie={loadingLatestMovie}
      errorLatest={errorLatestMovie}
      loadingFavList={loadingFavList}
      errorFavList={errorFavList}
      favorites={favorites}
      loadingFavs={loadingFavs}
      errorFavs={errorFavs}
      seen={seen}
      loadingSeen={loadingSeen}
      errorSeen={errorSeen}
    />
  );
};

export default Home;
