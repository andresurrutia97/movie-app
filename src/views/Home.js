import React from 'react';
import PropTypes from 'prop-types';

import Carousel from '../components/Ui/Carousel';
import LatestMovie from '../components/LatestMovie';
import Spinner from '../components/Ui/Spinner';
import Error from '../components/Ui/Error';
import { moviesShape, movieInfoShape, errorShape } from '../utils/propTypes';

const Home = ({
  popularMovies,
  loadingPopularMovies,
  errorPopular,
  topRatedMovies,
  loadingTopRatedMovies,
  errorTopRated,
  latestMovie,
  errorLatest,
  loadingLatestMovie,
  favorites,
  loadingFavs,
  errorFavs,
  seen,
  loadingSeen,
  errorSeen,
}) => {
  let latestMovieInfo = <Spinner />;
  let popularList = <Spinner />;
  let topRatedList = <Spinner />;
  let favoritesMovies = null;
  let seenMovies = null;

  if (latestMovie && !loadingPopularMovies) {
    latestMovieInfo = <LatestMovie movie={latestMovie} />;
  } else if (errorLatest) {
    latestMovieInfo = <Error />;
  }

  if (popularMovies && !loadingTopRatedMovies) {
    popularList = (
      <Carousel title="Popular">
        {popularMovies.map((movie) => (
          <Carousel.Item movie={movie} key={movie.id} />
        ))}
      </Carousel>
    );
  } else if (errorPopular) {
    popularList = <Error />;
  }

  if (topRatedMovies && !loadingLatestMovie) {
    topRatedList = (
      <Carousel title="Top Rated">
        {topRatedMovies.map((movie) => (
          <Carousel.Item movie={movie} key={movie.id} />
        ))}
      </Carousel>
    );
  } else if (errorTopRated) {
    topRatedList = <Error />;
  }

  if (favorites && !loadingFavs) {
    favoritesMovies = (
      <Carousel title="Favorites">
        {favorites.map((movie) => (
          <Carousel.Item movie={movie} key={movie.id} />
        ))}
      </Carousel>
    );
  } else if (errorFavs) {
    favoritesMovies = <Error />;
  }

  if (seen && !loadingSeen) {
    seenMovies = (
      <Carousel title="Seen">
        {seen.map((movie) => (
          <Carousel.Item movie={movie} key={movie.id} />
        ))}
      </Carousel>
    );
  } else if (errorSeen) {
    seenMovies = <Error />;
  }

  return (
    <>
      {latestMovieInfo}
      {favoritesMovies}
      {seenMovies}
      {popularList}
      {topRatedList}
    </>
  );
};

Home.propTypes = {
  popularMovies: PropTypes.arrayOf(PropTypes.shape(moviesShape)),
  topRatedMovies: PropTypes.arrayOf(PropTypes.shape(moviesShape)),
  favorites: PropTypes.arrayOf(PropTypes.shape(moviesShape)),
  seen: PropTypes.arrayOf(PropTypes.shape(moviesShape)),
  latestMovie: PropTypes.shape(movieInfoShape),
  errorLatest: PropTypes.shape(errorShape),
  errorPopular: PropTypes.shape(errorShape),
  errorTopRated: PropTypes.shape(errorShape),
  errorFavs: PropTypes.shape(errorShape),
  errorSeen: PropTypes.shape(errorShape),
  loadingPopularMovies: PropTypes.bool,
  loadingTopRatedMovies: PropTypes.bool,
  loadingLatestMovie: PropTypes.bool,
  loadingFavs: PropTypes.bool,
  loadingSeen: PropTypes.bool,
};

Home.defaultProps = {
  popularMovies: null,
  topRatedMovies: null,
  favorites: null,
  seen: null,
  latestMovie: null,
  errorLatest: null,
  errorPopular: null,
  errorTopRated: null,
  errorFavs: null,
  errorSeen: null,
  loadingPopularMovies: false,
  loadingTopRatedMovies: false,
  loadingLatestMovie: false,
  loadingFavs: false,
  loadingSeen: false,
};

export default Home;
