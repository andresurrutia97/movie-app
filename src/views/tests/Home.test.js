import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';

import Home from '../Home';
import { moviesData, movieInfo } from '../../utils/fakeData';
import Spinner from '../../components/Ui/Spinner';
import Carousel from '../../components/Ui/Carousel';
import LatestMovie from '../../components/LatestMovie';
import Error from '../../components/Ui/Error';

describe('<Home/>', () => {
  let wrapper;
  const initProps = {
    popularMovies: null,
    topRatedMovies: null,
    latestMovie: null,
    errorLatest: null,
    errorPopular: null,
    errorTopRated: null,
    loadingPopularMovies: false,
    loadingTopRatedMovies: false,
    loadingLatestMovie: false,
  };

  const component = (
    <BrowserRouter>
      <Home {...initProps} />
    </BrowserRouter>
  );

  beforeEach(() => {
    wrapper = mount(component);
  });

  it('should render correcty ', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should contain 3 instances of <Spinner/> while no info ', () => {
    expect(wrapper.find(Spinner)).toHaveLength(3);
  });

  it('should contain 2 instances of <Carousel/> if popularMovies and topRatedMovies props contain info  ', () => {
    wrapper.setProps({
      children: (
        <Home
          popularMovies={moviesData.results}
          topRatedMovies={moviesData.results}
        />
      ),
    });
    expect(wrapper.find(Carousel)).toHaveLength(2);
  });

  it('should contain an instance of <LatestMovie/> if latestMovie prop contains info  ', () => {
    wrapper.setProps({
      children: <Home latestMovie={movieInfo} />,
    });
    expect(wrapper.find(LatestMovie)).toHaveLength(1);
  });

  it('should contain 1 instance of <Error/> if errorLatest prop caintains info ', () => {
    wrapper.setProps({
      children: <Home errorLatest={{ error: 'error' }} />,
    });
    expect(wrapper.find(Error)).toHaveLength(1);
  });
});
