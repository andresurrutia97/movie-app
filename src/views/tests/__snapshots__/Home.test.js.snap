// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> should render correcty  1`] = `
Array [
  <div
    className="spinner"
  />,
  <div
    className="spinner"
  />,
  <div
    className="spinner"
  />,
]
`;
